File: rust_exercises/leetcode/Problem740/Solution.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        TreeMap<Integer, Integer> map = new TreeMap<>();
        for (int num: nums) {
            Integer prev = map.get(num);
            if (prev == null) {
                map.put(num, num);
            } else {
                map.put(num, prev.intValue() + num);
            }
        }
        
        int prev = 0;
        int sum0 = 0;
        int sum1 = 0;
        int count = 0;
        for (Map.Entry<Integer, Integer> entry: map.entrySet()) {
            if (count > 0 && entry.getKey() == prev + 1) {
                int temp0 = sum0;
                sum0 = Math.max(sum0, sum1);
                sum1 = temp0 + entry.getValue();
            } else {
                sum0 = Math.max(sum0, sum1);
                sum1 = sum0 + entry.getValue();
            }
            count += 1;
            prev = entry.getKey();
        }
        
        return Math.max(sum0, sum1);
    }
}

File: rust_exercises/leetcode/Problem740/solution.rs
use std::collections::BTreeMap;
use std::cmp::max;

impl Solution {
    pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
        let mut map = BTreeMap::new();
        for num in nums.into_iter() {
            if let Some(x) = map.get_mut(&num) {
                *x = *x + num;
            } else {
                map.insert(num, num);
            }
        }
        
        let mut sum0 = 0;
        let mut sum1 = 0;
        let mut prev = 0;
        
        for (i, (&k, &v)) in map.iter().enumerate() {
            if (i > 0 && k == prev + 1) {
                let temp = sum0;
                sum0 = max(sum0, sum1);
                sum1 = temp + v;
            } else {
                sum0 = max(sum0, sum1);
                sum1 = sum0 + v;
            }
            prev = k;
        }
        
        max(sum0, sum1)
    }
}
